src/elastio-snap.c

```diff
@@ -5395,53 +5559,70 @@ static inline unsigned long disable_page_protection(void) {
 static inline void reenable_page_protection(unsigned long cr0) {
 	wp_cr0(cr0);
 }
+#endif
 
-static void **find_sys_call_table(void){
-	long long offset;
-	void **sct;
-
-	if(!SYS_CALL_TABLE_ADDR || !SYS_MOUNT_ADDR || !SYS_UMOUNT_ADDR) return NULL;
-
-	offset = ((void *)kfree) - (void *)KFREE_ADDR;
-	sct = (void **)SYS_CALL_TABLE_ADDR + offset / sizeof(void **);
+/** generic function to manage system call table page permissions */
+static inline long syscall_mode_rw(void **syscall_table, int syscall_num)
+{
+#if defined(CONFIG_X86_64)
+	return disable_page_protection(); 
+#elif defined(CONFIG_ARM64)
+	return set_page_rw((unsigned long) (syscall_table + syscall_num));
+#else
+	return -1;
+#endif
+}
 
-	if(sct[__NR_mount] != (void **)SYS_MOUNT_ADDR + offset / sizeof(void **)) return NULL;
-	if(sct[__NR_umount2] != (void **)SYS_UMOUNT_ADDR + offset / sizeof(void **)) return NULL;
-#ifdef HAVE_SYS_OLDUMOUNT
-	if(sct[__NR_umount] != (void **)SYS_OLDUMOUNT_ADDR + offset / sizeof(void **)) return NULL;
+/** generic function to manage system call table page permissions */
+static inline long syscall_mode_ro(void **syscall_table, int syscall_num, unsigned long flags)
+{
+#if defined(CONFIG_X86_64)
+	reenable_page_protection(flags);
+#elif defined(CONFIG_ARM64)
+	return set_page_ro((unsigned long) (syscall_table + syscall_num));
+#else
+	return -1;
 #endif
+	return 0;
+}
 
-	LOG_DEBUG("system call table located at 0x%p", sct);
+static inline int syscall_set_hook(void **syscall_table,
+		int syscall_num, void **orig_hook, void *new_hook)
+{
+	long flags;
 
-	return sct;
-}
+	flags = syscall_mode_rw(syscall_table, syscall_num);
+	if (flags == -1)
+		return -EOPNOTSUPP;
```

Comment:
As far as I understood this idea properly, `disable_page_protection` for `X86_64` returns a flag necessary for `reenable_page_protection` call. And appropriate functions on `ARM64` don't use page protection state cr0 but can return an error. And you are trying to generalize these both things into one generic `syscall_mode_rw`.
I'd suggest to use additional parameter for the pointer to the flags variable and return just an error code and check it without potential confusion.

Reply:
Accepted.